core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -792,23 +797,29 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
   def flatMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2]): Stream[F2, O2] =
     Stream.fromFreeC[F2, O2](Algebra.uncons(get[F2, O]).flatMap {
       case Some((hd, tl)) =>
-        // nb: If tl is Pure, there's no need to propagate flatMap through the tail. Hence, we
-        // check if hd has only a single element, and if so, process it directly instead of folding.
-        // This allows recursive infinite streams of the form `def s: Stream[Pure,O] = Stream(o).flatMap { _ => s }`
-        val only: Option[O] = tl match {
-          case FreeC.Pure(_) => hd.head
-          case _             => None
-        }
-        only match {
-          case None =>
-            hd.map(f)
-              .foldRightLazy(Stream.fromFreeC(tl).flatMap(f))(_ ++ _)
-              .get
-
-          case Some(o) =>
-            f(o).get
+        tl match {
+          case FreeC.Result.Pure(_) if hd.size == 1  =>
+            // nb: If tl is Pure, there's no need to propagate flatMap through the tail. Hence, we
+            // check if hd has only a single element, and if so, process it directly instead of folding.
+            // This allows recursive infinite streams of the form `def s: Stream[Pure,O] = Stream(o).flatMap { _ => s }`
+            f(hd(0)).get
+
+          case _ =>
+            def go(rem: Chunk[O]): FreeC[Algebra[F2, O2, ?], Unit] =
+              rem.head match {
+                case Some(o) =>
+                  f(o).get.transformWith {
+                    case Result.Pure(_)   => go(rem.drop(1))
```

Comment:
I'd avoid using `rem.drop(1)` here as it will be expensive for large chunks. Instead, pass the index as a parameter to go and lookup the value via apply in each iteration.

Reply:
btw wouldn't apply(idx) be to slow for large Seq/List instances?

Reply:
No, b/c there are no such instances -- `Chunk` guarantees efficient indexed based access. The `Chunk.seq` constructor copies the seq in to a mutable buffer for that reason.

Reply:
cool